### PR TITLE
Wallet performance improvements numero uno

### PR DESCRIPTION
**Pull request by bokobza**
- Improved performance of when and how the lookup dictionary for addresses is loaded
- Added look-up on unspent outputs
- Changed from saving the wallet after every change to saving the wallet when
 1. a transaction is received from the network
 2. a block is done being processed and had some transactions affecting the wallet within it

- no benchmarking was done but the gain in speed is noticeable.

Kudos to @noescape00 for creating blocks with crazy transactions.
For reference, take a look at these blocks on Stratis Testnet:
333235
333445
333449
333946
334042
334044
334046
334048
334051
334052
334055
334056
334081
334084
334088
334094
334100

@dangershony

### CODE REVIEW

File: src/Stratis.Bitcoin.Features.Wallet/WalletManager.cs [hunk 1]
```diff
@@ -1194,23 +1181,38 @@ public void LoadKeysLookupLock()
...
+                            this.keysLookup[address.Pubkey] = address;
                     }
                 }
+            }
+        }
+
+        /// <summary>
+        /// Update the keys and transactions we're tracking in memory for faster lookups.
+        /// </summary>
+        public void UpdateKeysLookupLock(IEnumerable<HdAddress> addresses)
```
Comment (dangershony) [inline src/Stratis.Bitcoin.Features.Wallet/WalletManager.cs]: I would suggest (for the sake of avoiding double locks) to make this methods private `UpdateKeysLookupLock` `LoadKeysLookupLock` (if needed for unit testing then a public Internal method) and remove the lock from inside then make sure the callers are holding the lock, there are only 6 places where this method is called.

File: src/Stratis.Bitcoin.Features.Wallet/WalletManager.cs [hunk 2]
```diff
@@ -791,9 +794,13 @@ public bool ProcessTransaction(Transaction transaction, int? blockHeight = null,
                 }
 
                 // Check the inputs - include those that have a reference to a transaction containing one of our scripts and the same index.
-                foreach (TxIn input in transaction.Inputs.Where(txIn => this.keysLookup.Values.Distinct().SelectMany(v => v.Transactions).Any(trackedTx => trackedTx.Id == txIn.PrevOut.Hash && trackedTx.Index == txIn.PrevOut.N)))
+                foreach (TxIn input in transaction.Inputs)
                 {
-                    TransactionData tTx = this.keysLookup.Values.Distinct().SelectMany(v => v.Transactions).Single(trackedTx => trackedTx.Id == input.PrevOut.Hash && trackedTx.Index == input.PrevOut.N);
+                    this.outpointLookup.TryGetValue(input.PrevOut, out TransactionData tTx);
```
Comment (dangershony) [inline src/Stratis.Bitcoin.Features.Wallet/WalletManager.cs]: Why not?
```
if( !this.outpointLookup.TryGetValue(input.PrevOut, out TransactionData tTx)
{
   continue;
}
```

File: src/Stratis.Bitcoin.Features.Wallet/WalletManager.cs [hunk 3]
```diff
@@ -1215,6 +1221,12 @@ public void LoadKeysLookupLock()
                         this.keysLookup[address.ScriptPubKey] = address;
                         if (address.Pubkey != null)
                             this.keysLookup[address.Pubkey] = address;
+
+                        foreach (var unspentTransaction in address.UnspentTransactions())
+                        {
+                            this.outpointLookup[new OutPoint(unspentTransaction.Id, unspentTransaction.Index)] = unspentTransaction;
+                        }
+
```
Comment (dangershony) [inline src/Stratis.Bitcoin.Features.Wallet/WalletManager.cs]: space

File: src/Stratis.Bitcoin.Features.Wallet/WalletManager.cs (hunk 1 above)
Comment (dangershony) [reply src/Stratis.Bitcoin.Features.Wallet/WalletManager.cs]: Leave this to the wallet refactor when we make it stand alone I think this is good as it is now.

**Comment by dangershony** [comment]
I think this will also fix: #1228

File: src/Stratis.Bitcoin.Features.Wallet/WalletManager.cs (hunk 2 above)
Comment (bokobza) [reply src/Stratis.Bitcoin.Features.Wallet/WalletManager.cs]: yep

File: src/Stratis.Bitcoin.Features.Wallet/WalletManager.cs (hunk 3 above)
Comment (bokobza) [reply src/Stratis.Bitcoin.Features.Wallet/WalletManager.cs]: :space_invader:

**Comment by bokobza** [comment]
This is the first batch of improvements, more will come in the near future.

**Comment by noescape00** [comment]
Given that wallet is being updated every time it's changed do we need the async loop that saves it every 5 minutes? 
WalletManager.cs:177

**Comment by bokobza** [comment]
@noescape00 yeah this is to update the lastBlockSyncHeight and lastBlockSyncHash, even if there was no change in the wallet's addresses.

**Comment by noescape00** [comment]
Ah, makes perfect sense.

**Pull request merged**